### PR TITLE
feat(manager/pip-compile): Add lockedVersion to package file deps

### PR DESCRIPTION
**Pull request by not7cd**
## Changes

This adds `dep.lockedVersion` to every dependency in every package file. Locked versions are taken from a first lock file found. Further locked versions from other lock files are ignored.
A warning is raised if dep is found without a lock.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/pip-compile/extract.ts [hunk 1]
```diff
@@ -95,8 +99,7 @@ export async function extractAllPackageFiles(
         type: 'constraint',
       });
     }
-    // TODO(not7cd): handle locked deps
-    // const lockedDeps = extractRequirementsFile(content);
+    const lockedDeps = extractRequirementsFile(fileContent)!.deps;
```
Comment (viceice) [inline lib/modules/manager/pip-compile/extract.ts:102]:
```suggestion
    const lockedDeps = extractRequirementsFile(fileContent).deps;
```
why the assertion? should required or needs a check.

Comment (not7cd) [reply lib/modules/manager/pip-compile/extract.ts:102]: If we were able to successfully parse the header, then deps should exist. Although, there may be files with only header.

File: lib/modules/manager/pip-compile/extract.ts [hunk 2]
```diff
@@ -95,8 +96,14 @@ export async function extractAllPackageFiles(
         type: 'constraint',
       });
     }
-    // TODO(not7cd): handle locked deps
-    // const lockedDeps = extractRequirementsFile(content);
+    const lockedDeps = extractRequirementsFile(fileContent)?.deps;
+    if (!lockedDeps) {
+      logger.warn(
```
Comment (viceice) [inline lib/modules/manager/pip-compile/extract.ts:101]:
```suggestion
      logger.debug(
```

this should normally not happen and i think other managers are also silently ignore that

Comment (not7cd) [reply lib/modules/manager/pip-compile/extract.ts]: This should not happen especially after we successfully parsed the header.

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
Ups, conflicted 🙃

File: lib/modules/manager/pip-compile/extract.ts [hunk 3]
```diff
@@ -130,6 +137,20 @@ export async function extractAllPackageFiles(
...
+          const lockedVersion = lockedDeps?.find(
+            (lockedDep) =>
+              normalizeDepName(lockedDep.depName!) ===
+              normalizeDepName(dep.depName!),
+          )?.currentVersion;
+          if (lockedVersion) {
+            dep.lockedVersion = lockedVersion;
+          } else {
+            logger.warn(
+              `pip-compile: \`${dep.depName}\` not found in lock file \`${fileMatch}\``,
```
Comment (rarkins) [inline lib/modules/manager/pip-compile/extract.ts:150]:
```suggestion
              { depName: dep.depName, lockFile: fileMatch },
              `pip-compile: dependency not found in lock file`,
```

Avoid string interpolation in warning messages

File: lib/modules/manager/pip-compile/utils.ts [hunk 4]
```diff
@@ -17,3 +17,8 @@ export function generateMermaidGraph(
   });
   return `graph TD\n${lockFiles.join('\n')}\n${edges.join('\n')}`;
 }
+
+// https://packaging.python.org/en/latest/specifications/name-normalization/
+export function normalizeDepName(name: string): string {
+  return name.replace(/[-_.]+/g, '-').toLowerCase();
+}
```
[the remark this answers is not in the document]
Comment (not7cd) [reply lib/modules/manager/pip-compile/utils.ts:24]: Maybe, but the one in the datasource may not be correct. I also see that it is private?

**Review by rarkins — APPROVED** [review]

**Pull request merged**